File: sdk/python/dynaep/temporal/perception_profile.py
# ===========================================================================
# dynaep - Adaptive Perception Profiles (Python)
# Learns per-user temporal preferences from interaction patterns.
# Mirror of the TypeScript AdaptiveProfileManager.
# ===========================================================================

from __future__ import annotations

import json
import math
import time
from dataclasses import dataclass, field
from typing import Optional

from dynaep.temporal.perception_registry import PerceptionRegistry


# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------

@dataclass
class UserTemporalInteraction:
    user_id: str
    modality: str
    timestamp: float
    interaction_type: str  # response | interruption | replay_request | skip | slow_down_request | speed_up_request | completion | abandonment
    context_parameters: dict[str, float]
    response_latency_ms: Optional[float] = None


@dataclass
class ParameterAdjustment:
    parameter: str
    learned_offset: float
    sample_count: int
    variance: float
    last_updated_at: float


@dataclass
class ModalityPreference:
    modality: str
    parameter_adjustments: dict[str, ParameterAdjustment]
    confidence_score: float
    last_interaction_at: float


@dataclass
class AdaptivePerceptionProfile:
    user_id: str
    created_at: float
    updated_at: float
    interaction_count: int
    modalities: dict[str, ModalityPreference]


@dataclass
class AdaptiveProfileConfig:
    learning_rate: float
    erosion_half_life_ms: float
    min_samples_for_adjustment: int
    max_offset_from_comfortable: float
    forecast_enabled: bool
    persistence_enabled: bool
    persistence_path: str


# ---------------------------------------------------------------------------
# Signal Interpretation
# ---------------------------------------------------------------------------

@dataclass
class SignalResult:
    direction: int   # -1 = slow down, +1 = speed up, 0 = neutral
    magnitude: float  # 0.0 to 1.0


def _interpret_signal(interaction_type: str, response_latency_ms: Optional[float]) -> SignalResult:
    if interaction_type == "response":
        if response_latency_ms is None:
            return SignalResult(direction=0, magnitude=0.0)
        if response_latency_ms > 3000:
            return SignalResult(direction=-1, magnitude=0.6)
        if response_latency_ms > 1500:
            return SignalResult(direction=-1, magnitude=0.3)
        return SignalResult(direction=0, magnitude=0.0)
    elif interaction_type == "interruption":
        return SignalResult(direction=1, magnitude=0.5)
    elif interaction_type == "replay_request":
        return SignalResult(direction=-1, magnitude=0.7)
    elif interaction_type == "skip":
        return SignalResult(direction=1, magnitude=0.4)
    elif interaction_type == "slow_down_request":
        return SignalResult(direction=-1, magnitude=0.9)
    elif interaction_type == "speed_up_request":
        return SignalResult(direction=1, magnitude=0.9)
    elif interaction_type == "completion":
        return SignalResult(direction=0, magnitude=0.1)
    elif interaction_type == "abandonment":
        return SignalResult(direction=1, magnitude=0.6)
    else:
        return SignalResult(direction=0, magnitude=0.0)


# ---------------------------------------------------------------------------
# AdaptiveProfileManager
# ---------------------------------------------------------------------------

class AdaptiveProfileManager:
    """
    Manages per-user adaptive perception profiles that learn temporal
    preferences from interaction patterns. Profiles adjust annotation
    values within the comfortable range but NEVER exceed hard bounds.
    """

    def __init__(
        self,
        registry: PerceptionRegistry,
        config: AdaptiveProfileConfig,
    ) -> None:
        self._registry = registry
        self._config = config
        self._profiles: dict[str, AdaptivePerceptionProfile] = {}

    def ingest(self, interaction: UserTemporalInteraction) -> None:
        """Ingest a user interaction and update the corresponding profile."""
        user_id = interaction.user_id
        profile = self._profiles.get(user_id)

        if profile is None:
            profile = AdaptivePerceptionProfile(
                user_id=user_id,
                created_at=interaction.timestamp,
                updated_at=interaction.timestamp,
                interaction_count=0,
                modalities={},
            )
            self._profiles[user_id] = profile

        profile.interaction_count += 1
        profile.updated_at = interaction.timestamp

        pref = profile.modalities.get(interaction.modality)
        if pref is None:
            pref = ModalityPreference(
                modality=interaction.modality,
                parameter_adjustments={},
                confidence_score=0.0,
                last_interaction_at=interaction.timestamp,
            )
            profile.modalities[interaction.modality] = pref

        pref.last_interaction_at = interaction.timestamp

        signal = _interpret_signal(interaction.interaction_type, interaction.response_latency_ms)
        if signal.direction == 0 and signal.magnitude < 0.05:
            self._reinforce_neutral(pref, interaction.timestamp)
            self._update_confidence(pref)
            return

        modality_profile = self._registry.get_modality(interaction.modality)
        if modality_profile is None:
            return

        for param_name, bound in modality_profile.bounds.items():
            comfortable_width = bound.comfortable_max - bound.comfortable_min
            if comfortable_width <= 0:
                continue

            max_offset = comfortable_width * self._config.max_offset_from_comfortable
            signal_offset = signal.direction * signal.magnitude * max_offset

            adj = pref.parameter_adjustments.get(param_name)
            if adj is None:
                adj = ParameterAdjustment(
                    parameter=param_name,
                    learned_offset=0.0,
                    sample_count=0,
                    variance=0.0,
                    last_updated_at=interaction.timestamp,
                )
                pref.parameter_adjustments[param_name] = adj

            alpha = self._config.learning_rate
            old_offset = adj.learned_offset
            new_offset = alpha * signal_offset + (1 - alpha) * old_offset

            clamped_offset = max(-max_offset, min(max_offset, new_offset))

            delta = signal_offset - old_offset
            adj.variance = (1 - alpha) * adj.variance + alpha * delta * delta

            adj.learned_offset = clamped_offset
            adj.sample_count += 1
            adj.last_updated_at = interaction.timestamp

        self._update_confidence(pref)

    def _reinforce_neutral(self, pref: ModalityPreference, timestamp: float) -> None:
        """Decay existing offsets slightly toward zero on neutral signal."""
        decay_factor = 0.98
        for adj in pref.parameter_adjustments.values():
            adj.learned_offset *= decay_factor
            adj.sample_count += 1
            adj.last_updated_at = timestamp

    def _update_confidence(self, pref: ModalityPreference) -> None:
        """Recompute confidence score from sample counts."""
        total_samples = 0
        param_count = 0
        for adj in pref.parameter_adjustments.values():
            total_samples += adj.sample_count
            param_count += 1
        if param_count == 0:
            pref.confidence_score = 0.0
            return
        avg_samples = total_samples / param_count
        confidence = min(1.0, math.log2(avg_samples + 1) / 5.0)
        pref.confidence_score = confidence

    def get_profile(self, user_id: str) -> Optional[AdaptivePerceptionProfile]:
        """Return the adaptive profile for a user, or None."""
        return self._profiles.get(user_id)

    def adjust(
        self,
        user_id: str,
        modality: str,
        annotations: dict[str, object],
    ) -> dict[str, object]:
        """
        Apply the adaptive profile to annotations, returning adjusted
        values. Only adjusts if the profile has sufficient history.
        """
        profile = self._profiles.get(user_id)
        if profile is None:
            return dict(annotations)

        pref = profile.modalities.get(modality)
        if pref is None:
            return dict(annotations)

        modality_profile = self._registry.get_modality(modality)
        if modality_profile is None:
            return dict(annotations)

        adjusted = dict(annotations)

        for param_name, value in annotations.items():
            if not isinstance(value, (int, float)):
                continue

            adj = pref.parameter_adjustments.get(param_name)
            if adj is None or adj.sample_count < self._config.min_samples_for_adjustment:
                continue

            bound = modality_profile.bounds.get(param_name)
            if bound is None:
                continue

            adjusted_value = float(value) + adj.learned_offset
            clamped = max(bound.comfortable_min, min(bound.comfortable_max, adjusted_value))
            adjusted[param_name] = clamped

        return adjusted

    def erode_profiles(self) -> None:
        """Erode stale profile data based on the configured half-life."""
        now = time.time() * 1000
        half_life = self._config.erosion_half_life_ms

        for profile in self._profiles.values():
            for pref in profile.modalities.values():
                for adj in pref.parameter_adjustments.values():
                    elapsed = now - adj.last_updated_at
                    if elapsed <= 0:
                        continue
                    decay_factor = 0.5 ** (elapsed / half_life)
                    adj.learned_offset *= decay_factor

    def reset(self, user_id: str) -> None:
        """Reset a specific user's profile."""
        self._profiles.pop(user_id, None)

    def prune(self, retention_ms: float) -> int:
        """Prune profiles with no interactions within the retention window."""
        now = time.time() * 1000
        keys_to_delete = [
            uid for uid, profile in self._profiles.items()
            if (now - profile.updated_at) > retention_ms
        ]
        for key in keys_to_delete:
            del self._profiles[key]
        return len(keys_to_delete)

    def list_profiles(self) -> list[str]:
        """Return the list of all user IDs with active profiles."""
        return list(self._profiles.keys())

    def serialize(self) -> str:
        """Serialize all profiles to a JSON string."""
        data = {}
        for user_id, profile in self._profiles.items():
            modalities = {}
            for mod_name, pref in profile.modalities.items():
                adjustments = {}
                for pn, adj in pref.parameter_adjustments.items():
                    adjustments[pn] = {
                        "parameter": adj.parameter,
                        "learned_offset": adj.learned_offset,
                        "sample_count": adj.sample_count,
                        "variance": adj.variance,
                        "last_updated_at": adj.last_updated_at,
                    }
                modalities[mod_name] = {
                    "modality": pref.modality,
                    "parameter_adjustments": adjustments,
                    "confidence_score": pref.confidence_score,
                    "last_interaction_at": pref.last_interaction_at,
                }
            data[user_id] = {
                "user_id": profile.user_id,
                "created_at": profile.created_at,
                "updated_at": profile.updated_at,
                "interaction_count": profile.interaction_count,
                "modalities": modalities,
            }
        return json.dumps(data)

    def deserialize(self, data_str: str) -> None:
        """Load profiles from a previously serialized JSON string."""
        parsed = json.loads(data_str)
        self._profiles.clear()
        for user_id, pdata in parsed.items():
            modalities = {}
            for mod_name, mdata in pdata.get("modalities", {}).items():
                adjustments = {}
                for pn, adata in mdata.get("parameter_adjustments", {}).items():
                    adjustments[pn] = ParameterAdjustment(
                        parameter=adata["parameter"],
                        learned_offset=adata["learned_offset"],
                        sample_count=adata["sample_count"],
                        variance=adata["variance"],
                        last_updated_at=adata["last_updated_at"],
                    )
                modalities[mod_name] = ModalityPreference(
                    modality=mdata["modality"],
                    parameter_adjustments=adjustments,
                    confidence_score=mdata["confidence_score"],
                    last_interaction_at=mdata["last_interaction_at"],
                )
            self._profiles[user_id] = AdaptivePerceptionProfile(
                user_id=pdata["user_id"],
                created_at=pdata["created_at"],
                updated_at=pdata["updated_at"],
                interaction_count=pdata["interaction_count"],
                modalities=modalities,
            )

File: sdk/typescript/src/temporal/perception-profile.ts
// ===========================================================================
// @dynaep/core - Adaptive Perception Profiles
// Learns per-user temporal preferences from interaction patterns.
// Uses the TimesFM sidecar (when available) to forecast optimal timing
// for each user based on their response history.
// ===========================================================================

import type { PerceptionRegistry, PerceptionBounds, TemporalAnnotation } from "./perception-registry";
import type { ForecastSidecar } from "./forecast";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface UserTemporalInteraction {
  userId: string;
  modality: string;
  timestamp: number;
  interactionType: "response" | "interruption" | "replay_request" | "skip" |
    "slow_down_request" | "speed_up_request" | "completion" | "abandonment";
  contextParameters: Record<string, number>;
  responseLatencyMs: number | null;
}

export interface AdaptivePerceptionProfile {
  userId: string;
  createdAt: number;
  updatedAt: number;
  interactionCount: number;
  modalities: Record<string, ModalityPreference>;
}

export interface ModalityPreference {
  modality: string;
  parameterAdjustments: Record<string, ParameterAdjustment>;
  confidenceScore: number;
  lastInteractionAt: number;
}

export interface ParameterAdjustment {
  parameter: string;
  learnedOffset: number;
  sampleCount: number;
  variance: number;
  lastUpdatedAt: number;
}

export interface AdaptiveProfileConfig {
  learningRate: number;
  erosionHalfLifeMs: number;
  minSamplesForAdjustment: number;
  maxOffsetFromComfortable: number;
  forecastEnabled: boolean;
  persistenceEnabled: boolean;
  persistencePath: string;
}

// ---------------------------------------------------------------------------
// Signal Interpretation
// ---------------------------------------------------------------------------

interface SignalResult {
  direction: number;   // -1 = slow down, +1 = speed up, 0 = neutral
  magnitude: number;   // 0.0 to 1.0
}

function interpretSignal(interactionType: string, responseLatencyMs: number | null): SignalResult {
  switch (interactionType) {
    case "response": {
      if (responseLatencyMs === null) {
        return { direction: 0, magnitude: 0 };
      }
      // Long response latency suggests content was too fast
      if (responseLatencyMs > 3000) {
        return { direction: -1, magnitude: 0.6 };
      }
      if (responseLatencyMs > 1500) {
        return { direction: -1, magnitude: 0.3 };
      }
      // Quick response suggests pacing was comfortable
      return { direction: 0, magnitude: 0 };
    }
    case "interruption":
      return { direction: 1, magnitude: 0.5 };
    case "replay_request":
      return { direction: -1, magnitude: 0.7 };
    case "skip":
      return { direction: 1, magnitude: 0.4 };
    case "slow_down_request":
      return { direction: -1, magnitude: 0.9 };
    case "speed_up_request":
      return { direction: 1, magnitude: 0.9 };
    case "completion":
      return { direction: 0, magnitude: 0.1 };
    case "abandonment":
      return { direction: 1, magnitude: 0.6 };
    default:
      return { direction: 0, magnitude: 0 };
  }
}

// ---------------------------------------------------------------------------
// AdaptiveProfileManager
// ---------------------------------------------------------------------------

export class AdaptiveProfileManager {
  private readonly registry: PerceptionRegistry;
  private readonly forecastSidecar: ForecastSidecar | null;
  private readonly config: AdaptiveProfileConfig;
  private profiles: Map<string, AdaptivePerceptionProfile>;

  constructor(
    registry: PerceptionRegistry,
    forecastSidecar: ForecastSidecar | null,
    config: AdaptiveProfileConfig,
  ) {
    this.registry = registry;
    this.forecastSidecar = forecastSidecar;
    this.config = {
      learningRate: config.learningRate,
      erosionHalfLifeMs: config.erosionHalfLifeMs,
      minSamplesForAdjustment: config.minSamplesForAdjustment,
      maxOffsetFromComfortable: config.maxOffsetFromComfortable,
      forecastEnabled: config.forecastEnabled,
      persistenceEnabled: config.persistenceEnabled,
      persistencePath: config.persistencePath,
    };
    this.profiles = new Map();
  }

  /**
   * Ingest a user interaction and update the corresponding profile.
   * Creates a new profile if one does not exist for the user.
   */
  ingest(interaction: UserTemporalInteraction): void {
    const userId = interaction.userId;
    let profile = this.profiles.get(userId);

    if (!profile) {
      profile = {
        userId,
        createdAt: interaction.timestamp,
        updatedAt: interaction.timestamp,
        interactionCount: 0,
        modalities: {},
      };
      this.profiles.set(userId, profile);
    }

    profile.interactionCount += 1;
    profile.updatedAt = interaction.timestamp;

    let modalityPref = profile.modalities[interaction.modality];
    if (!modalityPref) {
      modalityPref = {
        modality: interaction.modality,
        parameterAdjustments: {},
        confidenceScore: 0,
        lastInteractionAt: interaction.timestamp,
      };
      profile.modalities[interaction.modality] = modalityPref;
    }

    modalityPref.lastInteractionAt = interaction.timestamp;

    const signal = interpretSignal(interaction.interactionType, interaction.responseLatencyMs);
    if (signal.direction === 0 && signal.magnitude < 0.05) {
      // Neutral signal, reinforce existing offsets slightly toward zero
      this.reinforceNeutral(modalityPref, interaction.timestamp);
      this.updateConfidence(modalityPref);
      return;
    }

    // Update parameter adjustments based on the signal
    const modalityProfile = this.registry.getModality(interaction.modality);
    if (!modalityProfile) {
      return;
    }

    for (const [paramName, bound] of Object.entries(modalityProfile.bounds)) {
      const comfortableWidth = bound.comfortable_max - bound.comfortable_min;
      if (comfortableWidth <= 0) {
        continue;
      }

      const maxOffset = comfortableWidth * this.config.maxOffsetFromComfortable;
      const signalOffset = signal.direction * signal.magnitude * maxOffset;

      let adj = modalityPref.parameterAdjustments[paramName];
      if (!adj) {
        adj = {
          parameter: paramName,
          learnedOffset: 0,
          sampleCount: 0,
          variance: 0,
          lastUpdatedAt: interaction.timestamp,
        };
        modalityPref.parameterAdjustments[paramName] = adj;
      }

      // Exponential moving average update
      const alpha = this.config.learningRate;
      const oldOffset = adj.learnedOffset;
      const newOffset = alpha * signalOffset + (1 - alpha) * oldOffset;

      // Clamp to max offset from comfortable midpoint
      const clampedOffset = Math.max(-maxOffset, Math.min(maxOffset, newOffset));

      // Update variance estimate
      const delta = signalOffset - oldOffset;
      adj.variance = (1 - alpha) * adj.variance + alpha * delta * delta;

      adj.learnedOffset = clampedOffset;
      adj.sampleCount += 1;
      adj.lastUpdatedAt = interaction.timestamp;
    }

    this.updateConfidence(modalityPref);
  }

  /**
   * Reinforce a neutral signal by slightly decaying existing offsets
   * toward zero, representing that the current pacing was acceptable.
   */
  private reinforceNeutral(pref: ModalityPreference, timestamp: number): void {
    const decayFactor = 0.98;
    for (const adj of Object.values(pref.parameterAdjustments)) {
      adj.learnedOffset *= decayFactor;
      adj.sampleCount += 1;
      adj.lastUpdatedAt = timestamp;
    }
  }

  /**
   * Recompute the confidence score for a modality preference based
   * on the total number of samples across all parameters.
   */
  private updateConfidence(pref: ModalityPreference): void {
    let totalSamples = 0;
    let paramCount = 0;
    for (const adj of Object.values(pref.parameterAdjustments)) {
      totalSamples += adj.sampleCount;
      paramCount += 1;
    }
    if (paramCount === 0) {
      pref.confidenceScore = 0;
      return;
    }
    const avgSamples = totalSamples / paramCount;
    // Confidence grows logarithmically with samples, capped at 1.0
    const confidence = Math.min(1.0, Math.log2(avgSamples + 1) / 5.0);
    pref.confidenceScore = confidence;
  }

  /**
   * Return the current adaptive profile for a user, or null if
   * no profile has been created.
   */
  getProfile(userId: string): AdaptivePerceptionProfile | null {
    const profile = this.profiles.get(userId);
    if (!profile) {
      return null;
    }
    return profile;
  }

  /**
   * Apply the adaptive profile to annotations, returning adjusted
   * values. Only adjusts if the profile has sufficient interaction
   * history (minSamplesForAdjustment).
   */
  adjust(
    userId: string,
    modality: string,
    annotations: TemporalAnnotation,
  ): TemporalAnnotation {
    const profile = this.profiles.get(userId);
    if (!profile) {
      return { ...annotations };
    }

    const pref = profile.modalities[modality];
    if (!pref) {
      return { ...annotations };
    }

    const modalityProfile = this.registry.getModality(modality);
    if (!modalityProfile) {
      return { ...annotations };
    }

    const adjusted: TemporalAnnotation = { ...annotations };

    for (const [paramName, value] of Object.entries(annotations)) {
      if (typeof value !== "number") {
        continue;
      }

      const adj = pref.parameterAdjustments[paramName];
      if (!adj || adj.sampleCount < this.config.minSamplesForAdjustment) {
        continue;
      }

      const bound = modalityProfile.bounds[paramName];
      if (!bound) {
        continue;
      }

      const adjustedValue = value + adj.learnedOffset;
      // Clamp within comfortable range (adaptive NEVER exceeds hard bounds)
      const clamped = Math.max(
        bound.comfortable_min,
        Math.min(bound.comfortable_max, adjustedValue),
      );
      adjusted[paramName] = clamped;
    }

    return adjusted;
  }

  /**
   * Use TimesFM to forecast optimal timing for the next interaction.
   * Returns null when TimesFM is unavailable or forecast is disabled.
   */
  async forecastOptimal(
    userId: string,
    modality: string,
  ): Promise<TemporalAnnotation | null> {
    if (!this.config.forecastEnabled || !this.forecastSidecar) {
      return null;
    }

    const isAvailable = await this.forecastSidecar.available();
    if (!isAvailable) {
      return null;
    }

    // Build a time series from the user's interaction history
    const profile = this.profiles.get(userId);
    if (!profile) {
      return null;
    }

    const pref = profile.modalities[modality];
    if (!pref) {
      return null;
    }

    const modalityProfile = this.registry.getModality(modality);
    if (!modalityProfile) {
      return null;
    }

    // Construct optimal annotations from the comfortable midpoint + learned offset
    const optimal: TemporalAnnotation = {};
    for (const [paramName, bound] of Object.entries(modalityProfile.bounds)) {
      const midpoint = (bound.comfortable_min + bound.comfortable_max) / 2;
      const adj = pref.parameterAdjustments[paramName];
      const offset = adj ? adj.learnedOffset : 0;
      const optimalValue = Math.max(
        bound.comfortable_min,
        Math.min(bound.comfortable_max, midpoint + offset),
      );
      optimal[paramName] = optimalValue;
    }

    return optimal;
  }

  /**
   * Erode stale profile data based on the configured half-life.
   * Offsets decay exponentially toward zero over time.
   */
  erodeProfiles(): void {
    const now = Date.now();
    const halfLife = this.config.erosionHalfLifeMs;

    for (const profile of this.profiles.values()) {
      for (const pref of Object.values(profile.modalities)) {
        for (const adj of Object.values(pref.parameterAdjustments)) {
          const elapsed = now - adj.lastUpdatedAt;
          if (elapsed <= 0) {
            continue;
          }
          const decayFactor = Math.pow(0.5, elapsed / halfLife);
          adj.learnedOffset *= decayFactor;
        }
      }
    }
  }

  /**
   * Reset a specific user's profile, removing all learned preferences.
   */
  reset(userId: string): void {
    this.profiles.delete(userId);
  }

  /**
   * Prune profiles with no interactions within the retention window.
   * Returns the number of profiles removed.
   */
  prune(retentionMs: number): number {
    const now = Date.now();
    let removed = 0;
    const keysToDelete: string[] = [];

    for (const [userId, profile] of this.profiles.entries()) {
      const age = now - profile.updatedAt;
      if (age > retentionMs) {
        keysToDelete.push(userId);
      }
    }

    for (const key of keysToDelete) {
      this.profiles.delete(key);
      removed += 1;
    }

    return removed;
  }

  /**
   * Serialize all profiles to a JSON string for persistence.
   */
  serialize(): string {
    const data: Record<string, AdaptivePerceptionProfile> = {};
    for (const [userId, profile] of this.profiles.entries()) {
      data[userId] = profile;
    }
    const serialized = JSON.stringify(data);
    return serialized;
  }

  /**
   * Load profiles from a previously serialized JSON string.
   */
  deserialize(data: string): void {
    const parsed = JSON.parse(data) as Record<string, AdaptivePerceptionProfile>;
    this.profiles.clear();
    for (const [userId, profile] of Object.entries(parsed)) {
      this.profiles.set(userId, profile);
    }
  }

  /**
   * Return the list of all user IDs with active profiles.
   */
  listProfiles(): string[] {
    const ids: string[] = [];
    for (const key of this.profiles.keys()) {
      ids.push(key);
    }
    return ids;
  }
}
